Drop voxels

paolafer opened the pull request:
In this PR a couple of functions are added to paolina, which allow one to eliminate voxels from a track, with energy lower than a threshold, and redistribute their energy among the closest voxels.
The function `drop_end_point_voxels` modifies the `voxels` argument.
I would require a review from @ausonandres, because he knows the algorithm, but he hasn't taken part in this particular implementation, and @mmkekic for an outsider view.

paolafer committed:
Add functions to drop voxels in tracks.

paolafer committed:
Add tests to functions and test file for a specific event.

jacg commented on invisible_cities/reco/paolina_functions_test.py:
```diff
@@ -469,4 +475,47 @@ def test_voxel_is_dropped(hits, requested_voxel_dimensions):
+    tot_initial_energy = sum(h.E for h in hits)
+    voxels = voxelize_hits(hits, requested_voxel_dimensions, strict_voxel_size=False)
+    energies = [v.E for v in voxels]
+    e_thr = random.uniform(min(energies), max(energies))
```
Generating random data inside Hypothesis properties (and keeping Hypothesis out of the loop) makes test failures irreproducible and therefore destroys Hypothesis' ability to shrink counterexamples, and present you with the simplest failing case. You really don't want to do this! Use `hypothesis.strategies.random_module` or `hypothesis.strategies.randoms`.

paolafer replied:
Done.

paolafer committed:
Eliminate random import and use hypothesis strategies instead.

paolafer committed:
Put small function in the bigger one to modify the argument.

jacg commented on invisible_cities/reco/paolina_functions.py:
```diff
@@ -222,60 +223,60 @@ def make_tracks(evt_number       : float,
+        min_v.energy += the_vox.E
+
+
+    mod_voxels = copy.deepcopy(voxels)
```
Why does a shallow copy not suffice?

mmkekic replied:
Because voxel contains hits and she is changing energies of the hits, so shallow copy wont protect us from keeping the hits of the original voxel list the same. On the other hand I dont quite like this new style, I think that drop voxels should be a function as it was in the original proposal.

jacg replied:
I suspected it had something to do with mutating the voxels, but then I was surprised to find that no tests fail if a shallow copy is performed instead.

Please provide a test that demonstrates the need for `deepcopy`, otherwise someone in the future will remove it as superfluous. Although if *none* of our extensive tests spot the problem ...

mmkekic replied:
Yes, my suggestion is to assert that the sum of hit energies for all hits inside the track before and after dropping voxels doesnt change. That fails with the shallow copy.
I would also suggest to make E  of a voxel a property of voxel that is being calculated based on hits contained inside, that way the voxel energy will be correctly updated when the hits energies inside the voxel change.

paolafer replied:
1. We can add a test that ensures that the original list of hits is not changed by the function.
2. The E property of a voxel was originally kept independent on the sum of the hit energies on purpose as a check that the voxelization algorithm (`voxelize_hits`) was performing correctly. I would like to keep this check, in case the voxelization algorithm is changed, but maybe there's another way to do that.

jjgomezcadenas replied:
> We can add a test that ensures that the original list of hits is not changed by the function.

We could do this by construction, passing a tuple rather than a list.

> The E property of a voxel was originally kept independent on the sum of the hit energies on purpose as a check that the voxelization algorithm (voxelize_hits) was performing correctly. I would like to keep this check, in case the voxelization algorithm is changed, but maybe there's another way to do that.

Noting prevents us from adding a second number, say E_voxel, E_sumHits

paolafer replied:
I put the `drop_voxel` function inside the `drop_end_point_voxels`function, in order to keep using a mutable object. I think that this is a case where mutable objects are useful, since it is a recursive algorithm (it keeps dropping voxels until a condition is reached). The alternative is to make a lot of useless copies, I think. Since there is no current user case for the `drop_voxel` function alone, I think we can implement it this way, and if in the future somebody asks for it, we can change the code.

paolafer replied:
As for the need of a `deepcopy`: I'm using it because we decided that functions in IC shouldn't have side effects, therefore we don't want the function to modify its argument. I will add a test to ensure that the function does not modify the input voxels, but I cannot figure out a test that demonstrates the need for not modifying the argument.

jacg replied:
The question was not why you copy, the question was why the copy needs to be *deep*. The answer is twofold:

1. The elements are mutable. For immutable elements you only ever need a shallow copy.

2. The elements are mutated. If they were never mutated, you could get away with a shallow copy.

> I put the `drop_voxel` function inside the `drop_end_point_voxels` function, in order to keep using a mutable object.

More precisely: in order not to violate the IC principle that functions should not modify their inputs. `drop_voxel` mutates its input, so by making it an *internal* implementation detail of `drop_end_point_voxels`, nobody can call it and be nastily surprised.

Let's introduce a bit of nomenclature: functions which mutate their inputs are called *destructive*, while ones which do not, are called *non-destructive*. So, we want IC functions to be non-destructive wherever pragmatically possible. But what if we do end up needing to write a destructive function?

I really like the Scheme approach to this: any function whose name ends in an exclamation mark (`!`) is destructive. Frequently one finds two different versions of the same function (for example `foo!` and `foo`) one destructive and one non-destructive. The former exists because it is more efficient than the latter, the latter exists because it is safer than the former. The universal and concise conventional meaning of `!` minimizes nasty surprises.

Unfortunately `!` is not a valid constituent character in Python, so we can't adopt that exact convention. But perhaps we should consider formalizing some such convention in IC.

Anyway, I think that making `drop_voxel` internal to `drop_end_point_voxels` is a decent idea. If we discover the need to pull it out, we should use whatever convention we agree on, to make it clear that it is a destructive function.

jjgomezcadenas replied:
BTW, Julia copies the same notation from Scheme! We could implement the convention
_D to mark destructive functions. But I also like the solution proposed by@paolafer!

jacg replied:
> Julia copies the same notation from Scheme! 

Unfortunately Julia is repeatedly failing to copy the similar `?` notation for predicates.

`_D` is probably the best we're going to come up with in Python. It's not bad: not too long but slightly ugly, and the ugliness is a good thing, as we want to feel a little discomfort when using destructive functions.

jacg commented on invisible_cities/reco/paolina_functions_test.py:
```diff
@@ -469,4 +475,47 @@ def test_voxel_is_dropped(hits, requested_voxel_dimensions):
+
+
+@given(bunch_of_hits, box_sizes, min_n_of_voxels)
+def test_total_energy_is_conserved(hits, requested_voxel_dimensions, min_voxels):
```
This test name should contain 'drop' somewhere.

paolafer replied:
Done.

jacg commented on invisible_cities/reco/paolina_functions.py:
```diff
@@ -219,3 +220,61 @@ def make_tracks(evt_number       : float,
+        return
+
+    pos = [h.pos for h in the_vox.hits]
+    qs = [h.E for h in the_vox.hits]
```
Please align the listcomps.

paolafer replied:
Done.

jacg commented on invisible_cities/reco/paolina_functions.py:
```diff
@@ -219,3 +220,61 @@ def make_tracks(evt_number       : float,
+
+    ### add voxel energy to hit and to voxel, separately
+    min_hit.energy += the_vox.E
+    min_v.energy += the_vox.E
```
Please align these two statements, starting at the first dot.

paolafer replied:
Done.

paolafer committed:
Cosmetics.

paolafer committed:
Add test to show the need of a deep copy of voxels in drop function.

paolafer committed:
Add new energy property for voxels, calculated from hits energy.

paolafer committed:
Cosmetics.

jacg commented on invisible_cities/reco/paolina_functions.py:
```diff
@@ -236,7 +236,7 @@ def drop_voxel(voxels: Sequence[Voxel], the_vox: Voxel):
 
         pos = [h.pos for h in the_vox.hits]
-        qs = [h.E for h in the_vox.hits]
+        qs  = [h.E for h in the_vox.hits]
```
The contents are still misaligned!

paolafer replied:
You're right, I must be alignment blind or something :-D.

paolafer committed:
Align stuff.

jacg committed:
Make test a bit more reader-friendly

When first reading this test, it took me too much time to figure out
what was going on, because the first 9 lines are pure noise. So the
lines have been grouped together under headings which should help
direct the reader's eye.

paolafer committed:
Add control on number of voxels within drop_voxel function.

paolafer committed:
Add test to show the behaviour for short tracks.

mmkekic commented on invisible_cities/reco/paolina_functions_test.py:
```diff
@@ -443,4 +451,89 @@ def test_contiguity(proximity, contiguity, are_neighbours):
+    final_trk_energies = [sum(vox.E for vox in t.nodes()) for t in final_trks]
+
+    assert tot_initial_energy == approx(tot_final_energy)
+    assert np.allclose(ini_trk_energies, final_trk_energies)
```
the test failed because the track order is not the same, you can either sort the energies when comparing them, or sort the tracks.

paolafer committed:
Order lists to be compared.

ausonandres reviewed (approved):
I've tested the functions and they work properly. So I approve this PR.
